src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,448 @@
+            if (operation.NextVariables.Length > 1)
+            {
+                return null;
+            }
+
+            if (!operation.NextVariables.IsEmpty)
+            {
+                var nextVariable = operation.NextVariables[0] as ILocalReferenceOperation;
+                if (nextVariable == null || nextVariable.Local?.Equals(foreachVariable) == false)
+                {
+                    // we do not support anything else than local reference for next variable
+                    // operation
+                    return null;
+                }
+            }
+
+            if (CheckIfForEachVariableIsWrittenInside(model, foreachVariable, foreachStatement))
+            {
+                return null;
+            }
+
+            var foreachCollection = RemoveImplicitConversion(operation.Collection);
+            if (foreachCollection == null)
+            {
+                return null;
+            }
+
+            GetInterfaceInfo(semanticFact, model, foreachVariable, foreachCollection,
+                out var explicitCastInterface, out var collectionNameSuggestion, out var countName);
+            if (countName == null)
+            {
+                return null;
+            }
+
+            var requireCollectionStatement = CheckRequireCollectionStatement(foreachCollection);
+            return new ForEachInfo(collectionNameSuggestion, countName, explicitCastInterface, foreachVariable.Type, requireCollectionStatement, foreachStatement);
+        }
+
+        private static void GetInterfaceInfo(
+            ISemanticFactsService semanticFact, SemanticModel model, ILocalSymbol foreachVariable, IOperation foreachCollection,
+            out ITypeSymbol explicitCastInterface, out string collectionNameSuggestion, out string countName)
+        {
+            explicitCastInterface = default;
+            collectionNameSuggestion = default;
+            countName = default;
+
+            // go through list of types and interfaces to find out right set;
+            var foreachType = foreachVariable.Type;
+            if (IsNullOrErrorType(foreachType))
+            {
+                return;
+            }
+
+            var collectionType = foreachCollection.Type;
+            if (IsNullOrErrorType(collectionType))
+            {
+                return;
+            }
+
+            // go through explicit types first.
+
+            // check array case
+            if (collectionType is IArrayTypeSymbol array && array.Rank == 1)
+            {
+                if (!IsExchangable(semanticFact, array.ElementType, foreachType, model.Compilation))
+                {
+                    return;
+                }
+
+                collectionNameSuggestion = "array";
+                explicitCastInterface = null;
+                countName = Length;
+                return;
+            }
+
+            // check string case
+            if (collectionType.Equals(model.Compilation.GetSpecialType(SpecialType.System_String)))
```

Comment:
You can directly check `collectionType.SpecialType == SpecialType.System_String`?

Reply:
yep.